File: go/shared/kafka/consumer.go
package kafka

import (
	"encoding/json"
	"fmt"
	"os"
	"os/signal"
	"strings"
	"syscall"

	"github.com/NCSU-Microservice-Benchmarking/vision_middleware/go/shared/types"
	"github.com/confluentinc/confluent-kafka-go/kafka"
)

type Consumer struct {
	Name             string
	Consumer         *kafka.Consumer
	Topic            string
	RequestProcessor types.RequestProcessor
	ProducerCallback types.ProducerCallback
	ProducerTopic    string
}

func NewConsumer(name string, options types.KafkaOptions, producerCallback types.ProducerCallback) (*Consumer, error) {
	configMap := &kafka.ConfigMap{
		"bootstrap.servers":    strings.Join(options.Brokers, ","),
		"group.id":             options.GroupID,
		"client.id":            options.ClientID,
		"auto.offset.reset":    "earliest",
		"enable.auto.commit":   false,
		"max.poll.interval.ms": 600000,
		"session.timeout.ms":   6000,
	}

	consumer, err := kafka.NewConsumer(configMap)
	if err != nil {
		return nil, err
	}

	return &Consumer{
		Name:             name,
		Consumer:         consumer,
		Topic:            options.Topics.Consumer,
		RequestProcessor: options.RequestProcessor,
		ProducerCallback: producerCallback,
		ProducerTopic:    options.Topics.Producer,
	}, nil
}

func (c *Consumer) Start() error {
	sigchan := make(chan os.Signal, 1)
	signal.Notify(sigchan, syscall.SIGINT, syscall.SIGTERM)

	// Subscribe to topics and consume
	if err := c.Subscribe([]string{c.Topic}); err != nil {
		return fmt.Errorf("error subscribing to topics: %v", err)
	}
	go c.Consume()
	fmt.Printf("%s consumer connected.", c.Name)
	return nil
}

func (c *Consumer) handleMessage(msg *kafka.Message) {
	// Parse message for the request
	var req types.Request
	json.Unmarshal(msg.Value, &req)

	// Process the request
	cached, newRequest, err := c.RequestProcessor(req)
	if err != nil {
		fmt.Println("Error processing request:", err)
		return
	}

	// Produce the message to the producer topic if it wasn't just handled by cache (that would return true)
	if c.ProducerTopic != "" && !cached {
		err := c.ProducerCallback(newRequest, c.ProducerTopic)
		if err != nil {
			fmt.Println("Error producing message:", err)
		}
	}
}

func (c *Consumer) Subscribe(topics []string) error {
	err := c.Consumer.SubscribeTopics(topics, nil)
	if err != nil {
		return err
	}
	fmt.Printf("%s consumer subscribed to topic '%s'.", c.Name, c.Topic)
	return nil
}

func (c *Consumer) Consume() {
	sigchan := make(chan os.Signal, 1)
	signal.Notify(sigchan, syscall.SIGINT, syscall.SIGTERM)

	run := true

	for run {
		select {
		case sig := <-sigchan:
			fmt.Printf("Caught signal %v: terminating\n", sig)
			run = false
		default:
			msg, err := c.Consumer.ReadMessage(-1)
			if err == nil {
				fmt.Printf("Received message: %v\n", string(msg.Value))
				go c.handleMessage(msg)
			} else {
				fmt.Printf("Error reading message: %v\n", err)
			}
		}
	}
}

func (c *Consumer) CommitOffsets() {
	c.Consumer.Commit()
}

func (c *Consumer) Shutdown() error {
	c.Consumer.Close()
	return nil
}

File: node/shared/kafka/consumer_.ts
import { Service as Microservice, kafkaOptions } from '../types/service';
import { request, requestProcessor } from '../types/request';
import { Consumer as KafkaConsumer, ConsumerSubscribeTopics, EachBatchPayload, Kafka, EachMessagePayload, ConsumerConfig } from 'kafkajs'

class Consumer implements Microservice.Consumer {
  
  private name: string;
  private consumer: KafkaConsumer;
  private topic: string;
  private requestProcessor: requestProcessor;

  private producerCallback?: (topic: string, message: any) => Promise<void>;
  private producerTopic?: string;

  constructor(name: string, options: kafkaOptions, producerCallback: any) {
    const { topics, requestProcessor } = options;
    this.name = name;
    this.topic = topics.consumer;
    this.consumer = this.create(options);
    if (topics.producer) {
      this.producerCallback = producerCallback;
      this.producerTopic = topics.producer;
    }
    this.requestProcessor = requestProcessor;
  }

  public async start(): Promise<void> {
    try {
      await this.subscribe();
      await this.consumer.run({
        eachMessage: async (messagePayload: EachMessagePayload) => {
          await this.handleMessage(messagePayload);
          return;
        }
      });

    } catch (error) {
      console.log('Error: ', error)
    }
  }

  public async subscribe(): Promise<void> {
    try {
      const topic: ConsumerSubscribeTopics = {
        topics: [this.topic],
        fromBeginning: false
      }
      await this.consumer.subscribe(topic);
    } catch (error) {
      console.log('Error: ', error)
    }
  }

  private async handleMessage(messagePayload: EachMessagePayload): Promise<void> {
    try {

      // log message
      const { topic, partition, message } = messagePayload;
      const prefix = `${topic}[${partition} | ${message.offset}] / ${message.timestamp}`
      console.log(`- ${prefix} ${message.key}#${message.value}`);

      // parse message for params
      const request: request = JSON.parse(message.value!.toString());

      // get a response
      const response: boolean | request = await this.requestProcessor(request);

      // produce the message to the correct topic if it wasn't just handled by cache (that would return true)
      response && this.producerTopic && response !== true &&
        await this.producerCallback!(this.producerTopic, response);
      return;

    } catch (error) {
      console.error('Error handling message:', error);
      return;
    }
  }

  public async shutdown(): Promise<void> {
    await this.consumer.disconnect();
  }

  private addEventListeners(consumer: KafkaConsumer): KafkaConsumer {
    consumer.on('consumer.connect', async () => {
      console.log(`${this.name} consumer connected.`)
    })
    return consumer;
  }

  private create(options: kafkaOptions): KafkaConsumer {

    const { brokers, clientId, groupId } = options;
    
    const kafka = new Kafka({ 
      clientId: clientId,
      brokers: brokers
    });

    const config: ConsumerConfig = {
      groupId: groupId
    }

    const consumer = this.addEventListeners(kafka.consumer(config));
    
    return consumer;
  }

}

export default Consumer;